### PR TITLE
Limit the number of transactions in pending set

### PR DESCRIPTION
**Pull request by tomusdrw**
With large pools and in case the node is not `--force-sealing` the pending set was actually being computed in the sync thread (propagate transactions on timeout). That could cause a significant delay, since there was nothing in the cache and the pool was pretty huge, but for propagation we never needed more than 64 first transactions.

- This PR introduces `max_len` argument to avoid constructing too large pending sets.
- Additionally we introduce `UnorderedIterator` that can be computed faster than the priority one. This for now is only used for `parity_allTransactions` RPC, but in future could be used in places where we don't really care that much about strict ordering (RPC filters maybe)

CC @folsen 

Partially related: #8696

### CODE REVIEW

File: ethcore/sync/src/api.rs [hunk 1]
```diff
@@ -373,7 +377,9 @@ struct SyncProtocolHandler {
 impl NetworkProtocolHandler for SyncProtocolHandler {
 	fn initialize(&self, io: &NetworkContext) {
 		if io.subprotocol_name() != WARP_SYNC_PROTOCOL_ID {
-			io.register_timer(0, Duration::from_secs(1)).expect("Error registering sync timer");
+			io.register_timer(PEERS_TIMER, Duration::from_millis(700)).expect("Error registering sync timer");
```
Comment (jimpo) [inline ethcore/sync/src/api.rs]: s/sync/peers/ in the expect message

File: ethcore/sync/src/api.rs [hunk 2]
```diff
@@ -373,7 +377,9 @@ struct SyncProtocolHandler {
...
+			io.register_timer(SYNC_TIMER, Duration::from_millis(1100)).expect("Error registering sync timer");
+			io.register_timer(TX_TIMER, Duration::from_millis(1300)).expect("Error registering sync timer");
```
Comment (jimpo) [inline ethcore/sync/src/api.rs]: The expect message should be changed probably.

File: miner/src/pool/queue.rs [hunk 3]
```diff
@@ -261,15 +280,27 @@ impl TransactionQueue {
 			scoring::NonceAndGasPrice,
 			Listener,
 		>) -> T,
+	{
+		debug!(target: "txqueue", "Re-computing pending set for block: {}", block_number);
+		let _timer = ::trace_time::PerfTimer::new("pool::collect_pending");
```
Comment (jimpo) [inline miner/src/pool/queue.rs]: Any reason not to use the `trace_time!` macro?

File: miner/src/pool/queue.rs [hunk 4]
```diff
@@ -198,45 +209,53 @@ impl TransactionQueue {
...
 
 		// Double check after acquiring write lock
 		let mut cached_pending = self.cached_pending.write();
-		if let Some(pending) = cached_pending.pending(block_number, current_timestamp, nonce_cap.as_ref()) {
+		if let Some(pending) = cached_pending.pending(block_number, current_timestamp, nonce_cap.as_ref(), max_len) {
 			return pending;
 		}
 
-		let pending: Vec<_> = self.collect_pending(client, block_number, current_timestamp, nonce_cap, |i| i.collect());
+		// In case we don't have a cached set, but we don't care about order
```
Comment (jimpo) [inline miner/src/pool/queue.rs]: Might be nice to test this case, and assert unordered results and that the cache is left alone.

Comment (tomusdrw) [reply miner/src/pool/queue.rs]: Test added.

**Comment by 5chdn** [comment]
Does not compile.

```
   Compiling parity-rpc v1.12.0 (file:///builds/parity/parity/rpc)
error[E0050]: method `ready_transactions` has 2 parameters but the declaration in trait `ethcore::miner::MinerService::ready_transactions` has 4
   --> rpc/src/v1/tests/helpers/miner_service.rs:211:42
    |
211 |     fn ready_transactions<C>(&self, _chain: &C) -> Vec<Arc<VerifiedTransaction>> {
    |                                             ^^ expected 4 parameters, found 2
    |
    = note: `ready_transactions` from trait: `fn(&Self, &C, usize, ethcore::miner::PendingOrdering) -> std::vec::Vec<std::sync::Arc<miner::pool::VerifiedTransaction>>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0050`.
error: Could not compile `parity-rpc`.
```

**Comment by dvdplm** [comment]
> Does not compile.

Confirmed. And while the above is easy enough to fix (`fn ready_transactions<C>(&self, _chain: &C, max_len: usize, ordering: PendingOrdering) -> Vec<Arc<VerifiedTransaction>>`) there are other errors after it; they seem to stem from the `TestBlockChainClient` and `ChainNotificationHandler` not working well together anymore: the tests assume a `new_transcations()` method that is not available.

**Pull request merged**